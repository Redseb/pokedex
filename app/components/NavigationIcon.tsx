import * as React from "react"
import { StyleProp, TextStyle, View, ViewStyle } from "react-native"
import { observer } from "mobx-react-lite"
import { colors, typography } from "app/theme"
import {Icon, IconTypes} from './'

export interface NavigationIconProps {
  /**
   * Whether the icon is currently focused.
   */
  focused: boolean,
  /**
   * What icon to show.
   */
  icon: IconTypes
}

/**
 * An Icon used in the BottomTabBar component. Depending on focused, it will be colored differently.
 */
export const NavigationIcon = observer(function NavigationIcon(props: NavigationIconProps) {
  const { focused, icon } = props

  return (
    <View style={$container}>
      <Icon icon={icon} size={25} color={focused ? colors.activeNavigation : colors.inactiveNavigation}/>
    </View>
  )
})

const $container: ViewStyle = {
  justifyContent: "center",
}