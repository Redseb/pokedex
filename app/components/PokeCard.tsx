import React from "react"
import { StyleProp, TextStyle, View, ViewStyle, TouchableOpacity } from "react-native"
import { observer } from "mobx-react-lite"
import { Pokemon } from "../types"
import { $shadow, ACTIVE_OPACITY, colors, rounding, spacing } from "../theme"
import { AutoImage, Text } from "./"
import { padNumber } from "app/utils/helpersCommon"
import { AppScreenName, navigate } from "app/navigators"

const CARD_HEIGHT = 100
const CARD_WIDTH = 160

interface PokeCardProps {
  /**
   * An optional style override useful for padding & margin.
   */
  style?: StyleProp<ViewStyle>
  /**
   * The Pokemon to get the data from
   */
  pokemon: Pokemon
}

/**
 * A small card displaying a Pokemon sprite, name, and id.
 * The card is clickable and will navigate to the Pokemon's details screen.
 * The card's color is based on the Pokemon's type.
 */
export const PokeCard = observer(function PokeCard(props: PokeCardProps) {
  const { style, pokemon } = props

  const primaryColor = colors.pokemonTypes[pokemon.types[0]].primary
  const secondaryColor = colors.pokemonTypes[pokemon.types[0]].secondary
  const spriteUrl = pokemon.sprites.front_default

  const $styles = [$container, style, { backgroundColor: primaryColor }]

  const onPress = () => {
    navigate(AppScreenName.PokemonDetails, { pokemon })
  }

  return (
    <TouchableOpacity style={$styles} activeOpacity={ACTIVE_OPACITY} onPress={onPress}>
      <AutoImage source={{ uri: spriteUrl }} maxHeight={CARD_HEIGHT * 0.75} maxWidth={CARD_WIDTH} />
      <View style={[$nameContainer, { backgroundColor: secondaryColor }]}>
        <Text style={$text}>{`#${padNumber(pokemon.id)} ${pokemon.name}`}</Text>
      </View>
    </TouchableOpacity>
  )
})

const $container: ViewStyle = {
  height: CARD_HEIGHT,
  margin: spacing.xxs,
  justifyContent: "space-between",
  alignItems: "center",
  borderRadius: rounding.s,
  ...$shadow.boxShadow,
}

const $nameContainer: ViewStyle = {
  justifyContent: "center",
  alignItems: "center",
  width: "100%",
  borderBottomLeftRadius: rounding.s,
  borderBottomRightRadius: rounding.s,
}

const $text: TextStyle = {
  color: colors.palette.neutral100,
}
