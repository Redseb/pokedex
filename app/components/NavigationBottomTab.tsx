import { TabsIcons } from 'app/navigators';
import { $shadow, colors, rounding, spacing } from 'app/theme';
import React from 'react';
import { View, TouchableOpacity, ViewStyle } from 'react-native';
import { NavigationIcon } from './NavigationIcon';

export const NavigationBottomTab = ({ state, descriptors, navigation }) => {
  return (
    <View style={$tabBarContainer}>
      {state.routes.map((route, index) => {
        const { options } = descriptors[route.key];

        const isFocused = state.index === index;

        const onPress = () => {
          const event = navigation.emit({
            type: 'tabPress',
            target: route.key,
            canPreventDefault: true,
          });

          if (!isFocused && !event.defaultPrevented) {
            // The `merge: true` option makes sure that the params inside the tab screen are preserved
            navigation.navigate({ name: route.name, merge: true });
          }
        };

        const onLongPress = () => {
          navigation.emit({
            type: 'tabLongPress',
            target: route.key,
          });
        };

        return (
          <TouchableOpacity
            accessibilityRole="button"
            accessibilityState={isFocused ? { selected: true } : {}}
            accessibilityLabel={options.tabBarAccessibilityLabel}
            testID={options.tabBarTestID}
            onPress={onPress}
            onLongPress={onLongPress}
            style={$tabContainer}
            key={index}
          >
            <NavigationIcon focused={isFocused} icon={TabsIcons[route.name]} />
          </TouchableOpacity>
        );
      })}
    </View>
  );
}

const $tabBarContainer: ViewStyle = {
  flexDirection: 'row',
  justifyContent: 'space-evenly',
  backgroundColor: colors.palette.primary500,
  borderTopRightRadius: rounding.s,
  borderTopLeftRadius: rounding.s,
  ...$shadow
}

const $tabContainer: ViewStyle = {
  flex: 1,
  justifyContent: 'center',
  alignItems: 'center',
  padding: spacing.sm
}