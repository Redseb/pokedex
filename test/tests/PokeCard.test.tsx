import React from "react"
import { render } from "@testing-library/react-native"
import { PokeCard } from "../../app/components/PokeCard"
import { mockPokemon } from "./mockPokemon"

describe("PokeCard", () => {
  it("renders the correct Pokemon name and ID", () => {
    const { getByText } = render(<PokeCard pokemon={mockPokemon} />)
    expect(getByText("#0001 Bulbasaur")).toBeTruthy()
  })

  it("matches snapshot", () => {
    const { toJSON } = render(<PokeCard pokemon={mockPokemon} />)
    expect(toJSON()).toMatchSnapshot()
  })
})
