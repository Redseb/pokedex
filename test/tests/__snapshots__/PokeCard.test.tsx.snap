// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeCard matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#7AC74C",
      "borderRadius": 5,
      "elevation": 5,
      "height": 100,
      "justifyContent": "space-between",
      "margin": 4,
      "opacity": 1,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.25,
      "shadowRadius": 3.84,
    }
  }
>
  <RCTImageView
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    internal_analyticTag={null}
    resizeMode="cover"
    source={
      [
        {
          "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
        },
      ]
    }
    style={
      {
        "height": 75,
        "overflow": "hidden",
        "width": 75,
      }
    }
  />
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderBottomLeftRadius": 5,
          "borderBottomRightRadius": 5,
          "justifyContent": "center",
          "width": "100%",
        },
        {
          "backgroundColor": "#4E8C18",
        },
      ]
    }
  >
    <Text
      style={
        [
          {},
          [
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            {
              "fontFamily": "spaceGroteskRegular",
            },
            {
              "color": "#191015",
            },
          ],
          undefined,
          undefined,
          {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      #0001 Bulbasaur
    </Text>
  </View>
</View>
`;
